readme.md
# prismarine-viewer

Web based viewer for servers and bots

[![NPM version](https://img.shields.io/npm/v/prismarine-viewer.svg)](http://npmjs.com/package/prismarine-viewer)
[![Build Status](https://github.com/PrismarineJS/prismarine-viewer/workflows/CI/badge.svg)](https://github.com/PrismarineJS/prismarine-viewer/actions?query=workflow%3A%22CI%22)
[![Discord](https://img.shields.io/badge/chat-on%20discord-brightgreen.svg)](https://discord.gg/GsEFRM8)
[![Gitter](https://img.shields.io/badge/chat-on%20gitter-brightgreen.svg)](https://gitter.im/PrismarineJS/general)
[![Irc](https://img.shields.io/badge/chat-on%20irc-brightgreen.svg)](https://irc.gitter.im/)
[![Issue Hunt](https://github.com/BoostIO/issuehunt-materials/blob/master/v1/issuehunt-shield-v1.svg)](https://issuehunt.io/r/PrismarineJS/prismarine-viewer)

[![Try it on gitpod](https://img.shields.io/badge/try-on%20gitpod-brightgreen.svg)](https://gitpod.io/#https://github.com/PrismarineJS/prismarine-viewer)

[<img src="https://prismarine.js.org/prismarine-viewer/test_1.18.1.png" alt="viewer" width="300">](https://prismarine.js.org/prismarine-viewer/)

## Install

```bash
npm install prismarine-viewer
```

## Example

```js
const mineflayer = require('mineflayer');
const mineflayerViewer = require('prismarine-viewer').mineflayer;

const bot = mineflayer.createBot({
  username: 'Bot',
});

bot.once('spawn', () => {
  mineflayerViewer(bot, { port: 3000 }); // Start the viewing server on port 3000

  // Draw the path followed by the bot
  const path = [bot.entity.position.clone()];
  bot.on('move', () => {
    if (path[path.length - 1].distanceTo(bot.entity.position) > 1) {
      path.push(bot.entity.position.clone());
      bot.viewer.drawLine('path', path);
    }
  });
});
```

More examples:

- First person bot [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/firstperson_bot.js)
- Record view as video file [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/headless.js)
- Streaming video to a python script [example](https://github.com/PrismarineJS/prismarine-viewer/tree/master/examples/python)
- Visualize a world, without a bot [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/standalone.js)
- Visualize the world coming from a proxy [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/proxy.js)
- Click to move [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/clickmove.js)
- Use the core api for viewing worlds [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/core)
- Create an electron app with viewer [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/electron)
- Create a fully front end viewer with an in memory world [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/standalone)
- A minecraft web client example, using mineflayer and a websocket proxy [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/web_client)
- Export parts of worlds as screenshot or 3d models [example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/exporter)

## Projects using prismarine-viewer

- [prismarine-web-client](https://github.com/PrismarineJS/prismarine-web-client) A minecraft client in your browser

## API

### prismarine-viewer

#### viewer

The core rendering library. It provides Viewer and WorldView which together make it possible to render a minecraft world.
Check its [API](viewer/README.md)

#### mineflayer

Serve a webserver allowing to visualize the bot surrounding, in first or third person. Comes with drawing functionnalities.

```js
const { mineflayer } = require('prismarine-viewer');
```

Options:

- `viewDistance` view radius, in chunks, default: `6`
- `firstPerson` is the view first person ? default: `false`
- `port` the port for the webserver, default: `3000`

[example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/bot.js)

#### standalone

Serve a webserver allowing to visualize a world.

```js
const { standalone } = require('prismarine-viewer');
```

Options:

- `version` the version to use, default: `1.13.2`
- `generator` a world generator function, default: `(x, y, z) => 0`
- `center` a vec3 to center the view on, default: `new Vec3(0, 0, 0)`
- `viewDistance` view radius, in chunks, default: `6`
- `port` the port for the webserver, default: `3000`

[example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/standalone.js)

#### headless

Render the bot view and stream it to a file or over TCP.

```js
const { headless } = require('prismarine-viewer');
```

Options:

- `viewDistance` view radius, in chunks, default: `6`
- `output` the output file or a `host:port` address to stream to, default: `output.mp4`
- `frames` number of frames to record, `-1` for infinite, default: `200`
- `width` the width of a frame, default: `512`
- `height` the height of a frame, default: `512`

[example](https://github.com/PrismarineJS/prismarine-viewer/blob/master/examples/headless.js)

### Drawing (mineflayer mode)

All drawing function have a unique id that can be used to replace or erase the primitive.

#### bot.viewer.drawLine (id, points, color=0xff0000)

Draw a line passing through all the `points`.

#### bot.viewer.erase (id)

Remove the primitive with the given id from the display.

#### bot.viewer.close ()

Stop the server and disconnect users.

## Tests

`node_modules/.bin/jest --verbose --runInBand --forceExit -t "1.16.4"`


HISTORY.md
### 1.25.0

- Remove unused mc-data from bundle

### 1.24.0

- Fix threejs disposal
- Expose ambientLight and directionalLight

### 1.23.0

- Bumb Mineflayer to 4.x
- Bumb Mineflayer-pathfinder to 2.x
- Bumb fs-extra to 10.x
- Bumb minecraft-data to 3.x

### 1.22.0

- support 1.18.0

### 1.21.0

- Dependency bumb
- Pin three dependency
- Fixed a few memory leaks
- Add waitForChunksToRender() function

### 1.20.0

- 1.17.1 support

### 1.19.3

- Fix standalone

### 1.19.2

- biome cache bug fix

### 1.19.1

- Lighten ao slightly
- Fix some missing rotations
- Fix some entity texture paths

### 1.19.0

- Add tints (thanks @Moondarker)

### 1.18.2

- Use entity name correctly

### 1.18.1

- Fix self ghost entity

### 1.18.0

- Add names above players

### 1.17.0

- expose version in viewer

### 1.16.2

- properly fix water bug
- improve input perfs

### 1.16.1

- fix rendering bug

### 1.16.0

- Fix bugs
- Improve water rendering

### 1.15.1

- Improve entity rendering
- Fix entity name matching on older versions

### 1.15.0

- Interpolate movements
- Add entity models

### 1.14.0

- Add viewer.close()
- Use real entity height and width
- Add points primitive

### 1.13.0

- make init and updatePosition not blocking for a faster apparent rendering
- use production mode webpack

### 1.12.0

- add keyboard controls (@extremeheat)

### 1.11.1

- use iterators from pworld

### 1.11.0

- 1.16.4 support
- Fix rotations

### 1.10.0

- Add prefix option
- Paste schematic in standalone example

### 1.9.3

- electron path fix

### 1.9.0

- Refactor package to expose rendering api
- Add electron demo
- Add static standalone demo

### 1.8.0

- Fix dirty chunk unloading bug (thanks @extremeheat)
- Add blockClicked event and clickmove example

### 1.7.0

- Added headless rendering
- Added writing to video file example
- Added streaming to python example
- Fixed camera rotation

### 1.6.0

- Model rotations (thanks @iRath96)
- Handling transparent blocks better (thanks @iRath96)
- Implement multi-part block models (thanks @iRath96)
- Respects pixel device ratios (thanks @iRath96)
- Fixed UV coordinates for blocks like hoppers (thanks @iRath96)
- Liquids should be 14px instead of 16px high (thanks @iRath96)

### 1.5.0

- Cache blocks for a 30% performance increase
- Enable compression

### 1.4.0

- Use more than 1 worker thread for mesh generation

### 1.3.0

- stop using bot.\_columns
- first person vision
- Basic liquids implementation
- Fixed canvas CSS

### 1.2.2

- Fix changing dimensions (in mineflayer)
- Improve examples

### 1.2.1

- Protect diverse null/undefined values

### 1.2.0

- Dynamic loading/unloading of chunks

### 1.1.0

- Build mesh into a separate worker thread

### 1.0.2

- Use correct textures and models for most of the blocks

### 1.0.1

- Fix import of src/index.js

### 1.0.0

- Initial version, API is not 100% stable but it works


lib/mineflayer.js
const EventEmitter = require('events');
const { WorldView } = require('../viewer');

module.exports = (
  bot,
  { viewDistance = 6, firstPerson = false, port = 3000, prefix = '' },
) => {
  const express = require('express');

  const app = express();
  const http = require('http').createServer(app);

  const io = require('socket.io')(http, { path: prefix + '/socket.io' });

  const { setupRoutes } = require('./common');
  setupRoutes(app, prefix);

  const sockets = [];
  const primitives = {};

  bot.viewer = new EventEmitter();

  bot.viewer.erase = id => {
    delete primitives[id];
    for (const socket of sockets) {
      socket.emit('primitive', { id });
    }
  };

  bot.viewer.drawBoxGrid = (id, start, end, color = 'aqua') => {
    primitives[id] = { type: 'boxgrid', id, start, end, color };
    for (const socket of sockets) {
      socket.emit('primitive', primitives[id]);
    }
  };

  bot.viewer.drawLine = (id, points, color = 0xff0000) => {
    primitives[id] = { type: 'line', id, points, color };
    for (const socket of sockets) {
      socket.emit('primitive', primitives[id]);
    }
  };

  bot.viewer.drawPoints = (id, points, color = 0xff0000, size = 5) => {
    primitives[id] = { type: 'points', id, points, color, size };
    for (const socket of sockets) {
      socket.emit('primitive', primitives[id]);
    }
  };

  io.on('connection', socket => {
    socket.emit('version', bot.version);
    sockets.push(socket);

    const worldView = new WorldView(
      bot.world,
      viewDistance,
      bot.entity.position,
      socket,
    );
    worldView.init(bot.entity.position);

    worldView.on('blockClicked', (block, face, button) => {
      bot.viewer.emit('blockClicked', block, face, button);
    });

    for (const id in primitives) {
      socket.emit('primitive', primitives[id]);
    }

    function botPosition() {
      const packet = {
        pos: bot.entity.position,
        yaw: bot.entity.yaw,
        addMesh: true,
      };
      if (firstPerson) {
        packet.pitch = bot.entity.pitch;
      }
      socket.emit('position', packet);
      worldView.updatePosition(bot.entity.position);
    }

    bot.on('move', botPosition);
    worldView.listenToBot(bot);
    socket.on('disconnect', () => {
      bot.removeListener('move', botPosition);
      worldView.removeListenersFromBot(bot);
      sockets.splice(sockets.indexOf(socket), 1);
    });
  });

  http.listen(port, () => {
    console.log(`Prismarine viewer web server running on *:${port}`);
  });

  bot.viewer.close = () => {
    http.close();
    for (const socket of sockets) {
      socket.disconnect();
    }
  };
};
